### PR TITLE
instance container interface + inMem imp

### PR DESCRIPTION
**Pull request by nivBlox**
Add instance container interface and a inMemory implantation for the qbft controller to use.

### CODE REVIEW

File: qbft/instance_store.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+func (i *inMemContainer) AddNewInstance(instance *Instance) {
+	for idx := i.capacity - 1; idx > 0; idx-- {
+		i.container[idx] = i.container[idx-1]
+	}
+	i.container[0] = instance
+}
+
+// AddNewInstanceAtPosition adding instance at specific position in store
+func (i *inMemContainer) AddNewInstanceAtPosition(position int, instance *Instance) {
+	if position >= i.capacity {
```
Comment (amir-blox) [inline qbft/instance_store.go:66]: Check negative values

File: qbft/instance_store.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
+package qbft
+
+import "encoding/json"
+
+// InstanceContainer interface for managing multi instance lifecycle
+type InstanceContainer interface {
+	All() []*Instance
+	FindInstance(height Height) *Instance
+	FindInstanceByPosition(position int) *Instance
```
Comment (alonmuroch) [inline qbft/instance_store.go:9]: what is position?

File: qbft/controller.go [hunk 3]
```diff
@@ -58,7 +37,7 @@ func NewController(
 		Height:              -1, // as we bump the height when starting the first instance
 		Domain:              domain,
 		Share:               share,
-		StoredInstances:     InstanceContainer{},
+		StoredInstances:     NewInMemContainer(HistoricalInstanceCapacity),
```
Comment (alonmuroch) [inline qbft/controller.go:40]: It should receive the instance via the func params not hardcoded

File: qbft/testing_utils.go [hunk 4]
```diff
@@ -92,11 +92,18 @@ var testingInstanceStruct = &Instance{
...
 	},
 }
-var testingControllerStruct = &Controller{
-	Identifier: []byte{1, 2, 3, 4},
-	Height:     Height(1),
-	Share:      testingShare,
-	StoredInstances: [HistoricalInstanceCapacity]*Instance{
-		testingInstanceStruct,
-	},
+var testingControllerStruct = func() *Controller {
```
Comment (alonmuroch) [inline qbft/testing_utils.go:95]: who calls this?

File: qbft/instance_store.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
+package qbft
```
Comment (alonmuroch) [inline qbft/instance_store.go:1]: this should be in testing utils not here

File: qbft/controller.go [hunk 6]
```diff
@@ -227,14 +206,50 @@ func (c *Controller) Decode(data []byte) error {
...
 
+// UnmarshalJSON add a custom marshaling for Controller
+func (c *Controller) UnmarshalJSON(data []byte) error {
+	var objmap map[string]*json.RawMessage
+	if err := json.Unmarshal(data, &objmap); err != nil {
+		return err
+	}
+	fmt.Println(objmap)
+
+	var imc inMemContainer
```
Comment (alonmuroch) [inline qbft/controller.go:225]: That shouldn't be here.. spec is not implementation specific

**Review by alonmuroch — CHANGES_REQUESTED** [review]

**Comment by alonmuroch** [comment]
I don't think this approach is good... it complicates things on multiple fronts. 
Might better have the store outside of controller marshaling/ root struct or something but that will cause issues with tests

File: qbft/controller.go (hunk 3 above)
Comment (nivBlox) [reply qbft/controller.go:40]: just aligned the old code.. no change has made

File: qbft/testing_utils.go (hunk 4 above)
Comment (nivBlox) [reply qbft/testing_utils.go:95]: qbft/controller_test.go:72

File: qbft/instance_store.go (hunk 5 above)
Comment (nivBlox) [reply qbft/instance_store.go:1]: why testing? 
i can separate the interface and the implementation if that what you mean

File: qbft/controller.go (hunk 6 above)
Comment (nivBlox) [reply qbft/controller.go:225]: but it is implemented an in memory before. the other option is to not use interface

**Comment by nivBlox** [comment]
> that

the previews container is rooted in the code & tests 
without the marshaling the instance store will not pass to the encoded bytes

**Comment by nivBlox** [comment]
*DO NOT MERGE*

Use a different way to implement loading instance from storage

**Pull request closed without merging**